piano keys, tone generator parameters, and sprite blending

Kouzeru opened the pull request:
Adding a piano keys to make ease of determining pitch value of musical notes.
I also had corrected the parameters of the tone generator into which I consider much more appropriate.

preview: https://kouzeru.github.io/Octo/

Kouzeru committed:
piano keys, and tone generator parameters

JohnEarnest commented:
I disagree with a number of aspects of this design:

- showing the piano keys at the top is extremely confusing; the most important elements of the UI should be the pattern and the pitch. the piano keys are logically part of the tone generator.
- it is no longer possible to directly approximate a desired frequency; pitch and pattern are totally disconnected.

In addition, the piano keys themselves provide no audio feedback; it's awkward to have to hit "play" after every change

Kouzeru commented:
I'm putting the piano key on the middle-

And my interest is mostly into music, and I usually don't give mind to frequency values...
The only matter is frequency interval, only to grasp with given distance between one note to another.

Here's the only reason why I used number cycles (harmonics) to determine frequency of pattern in the tone generator.
Combining interval 4 : 5 : 6, creating a C major chord.

https://user-images.githubusercontent.com/46141631/133897202-40156d9e-4157-4c36-8efd-2d4b0eac5969.mp4

Kouzeru committed:
ui reorder, putting the piano to middle

Kouzeru committed:
Merge branch 'JohnEarnest:gh-pages' into piano-keys

Kouzeru committed:
frequency monitor below tone generator for nerds

Kouzeru commented:
I'm adding the frequency monitor to show the relation between frequency and pattern below the tone generator.

ephphatha commented:
I like the proposed inputs for the tone generator, having a duty % and target number of pulses is much more intuitive in my mind than needing to specify a bitwidth and how much of that should be active. Also allowing inverting and reversing the pattern makes it faster to do blends. And relabelling the Random button to Noise makes sense to me.

The bits I found confusing were the play buttons felt disconnected from the buffer that they actually use. Having play below the keyboard while it uses the buffer in Audio Pattern had me pressing it and getting no output till I worked out what the go was. I had a quick play around with moving controls and ended up with this UI which I thought made the relationship a bit clearer (moving the play buttons alongside the representation of the buffer they use):
![Octo Audio editor showing play buttons moved to the right of the visual buffer representations](https://user-images.githubusercontent.com/357684/133912472-10e5daa4-d3db-4250-9f78-6fae541093dd.png)

I also changed the position of the Pitch input in that screenshot but I'm not particularly happy with that. The aim there was to show that the keyboard visualisation is another method of selecting a pitch value so having those controls next to each other seems appropriate. On that note I prefer how this implementation behaves in that the selected pitch is applied to both the tone generator buffer and the audio pattern buffer. In the old editor changing the frequency changes the pitch used for the tone generator, but you can't see what the actual pitch value is until you click Generate, overwriting the active pattern. The relationship could be made clearer but I don't have the UX skills to know how to do that effectively.

The live frequency display is great, one of the things I found most confusing about the existing Octo audio editor is the relationship between the frequency input for the tone generator and the resulting pattern and pitch. This lets me see how changing the number of cycles and the pitch adjusts the tone (and also highlights that duty% is more like a volume control). The preview frequency doesn't handle out of range values nicely though as trying to set a cycle count higher than 64 shows an overly high frequency until you click play for the tone generator. This is an issue with the current implementation anyway, there's no input validation.

Kouzeru committed:
rearranging the UI, following the suggestion

Kouzeru committed:
boom, sprite blender, and stuff

Kouzeru commented:
Thank you for the lovely feedback! It's nice to know something I didn't even realize about being self-explanatory, I'm quickly applying the things that you had suggested here, and fixing what you issued about. :)

---

By the way, I'm looking forward for the blending stuff, now its introduced in Sprite Editor and I believe it'll very useful to hasten sprite creation. Also, the sprite invert, flip and rotate, since there was no easy way to do it.

![AUZXahBkp7](https://user-images.githubusercontent.com/46141631/133935109-2a09c674-62c9-470e-9266-fb03db398732.gif)

Kouzeru committed:
piano click play feedback, and some optimizations stuff

Kouzeru commented:
The piano key is now clickable to play pattern directly from it...

Left-click plays pattern from the pattern editor, right-click plays pattern from tone generator.
![image](https://user-images.githubusercontent.com/46141631/134010325-6c68ce1d-25d2-4471-8d7a-b51d405618cb.png)
( also thought forward for someone's who's lefty, hence I don't need to specify which and which )

As the side changes of adding the mouse-release to the handler on `drawOnCanvas()` in order to achieve this, I've done some optimizations upon the sprite and audio editors, which only refresh the text box when mouse is released. I had to do this because the DOM refresh is slow, hence editing sprite and audio was lagging.

The other minor changes is also readjusting the audioScriptProcessor buffer size, to make it dynamic depending on the environment instead of being stuck at some constant.

Kouzeru committed:
oops, correcting the generate was somewhat not playing

Kouzeru commented:
It seem lot of changes packed here 😓

But, I'm still wondering if there is any further thoughts-

I'm done with everything, feel free to make correction upon anything as I made some reviews there.

Kouzeru committed:
Merge branch 'gh-pages' into piano-keys

Kouzeru committed:
relabel "oscillations" to "pulse" & deobfuscate bufferSize chooser

Kouzeru commented on index.html:
```diff
@@ -990,10 +1018,9 @@ <h1>Touch Input Configuration</h1>
   var stepFrame = _ => {
-    diff = (Date.now() - last); last += diff;
-    for(var k = 0; origin < last - frameTime && k < 2; origin += frameTime, k++){
+    for(var k = 0; timeCount < Date.now() && k < 2; timeCount += frameTime, k++){
```
some simplifications, as these variables isn't needed. it was formerly used for debugging and i forgot to remove them

Kouzeru commented on js/shared.js:
```diff
@@ -206,7 +206,10 @@ function audioSetup(emulator) {
+		var bufferSize = // set bufferSize according to environment's samplerate
+			audio.sampleRate <  64000 ? 2048 : // for 48000hz or 44100hz or less
+			audio.sampleRate < 128000 ? 4096 : 8192; // for 96000hz or more
+		audioNode = audio.createScriptProcessor(bufferSize, 1, 1);
```
this determines an ideal bufferSize according to environment's samplerate in a goal to achieve low latency as possible while avoiding buffer underrun (which results some crackly interference).

As myself always switching these options:

![image](https://user-images.githubusercontent.com/46141631/134659508-78694859-b523-4e58-a615-45b847a7cf94.png)

I'm choosing 44100hz which meant for performance.

Kouzeru commented on index.html:
```diff
@@ -81,28 +79,59 @@
+          <div id='sprite-vflip'>&vArr;</div>
+          <div id='sprite-hflip'>&hArr;</div>
+          <div id='sprite-rotCC' style='display:none'>&#x21BA;</div>
+          <div id='sprite-rotCW' style='display:none'>&#x21BB;</div>
```
the rotation buttons are hidden by default when working with 8x15 sprite mode,
and only shown when working on 16x16 sprite mode

I just couldn't imagine to rotating sprite when width and height isn't equal,
I mean, it needs more complication than simply swapping x and y positions with the diagonal flip.

Kouzeru commented on index.html:
```diff
@@ -81,28 +79,59 @@
+          </span>
+          <span class='pair'>Pitch: <input id='audio-pitch' value='64'></span>
+        </div>
+        <center><i><div id='audio-error'>left / right click to play audio / tone pattern</div></i></center>
```
literally, the audio-error element being used as instruction label. And it'd immediately replaced with 'Your browser does not support HTML5 Audio!' when audio error occurs

![image](https://user-images.githubusercontent.com/46141631/134704382-6961c19f-4743-4306-89ec-d327429d327d.png)
_(hm, beautiful)_

Kouzeru commented on index.html:
```diff
@@ -112,14 +141,11 @@ <h1>Tone Generator</h1>
-          <div id='audio-tone-preview' class='button'>Play</div>
-          <div id='audio-generate'     class='button'>Generate</div>
+          <div class='pair'>Duty: <input id='tone-duty' value='50'> %</div>
+          <div class='pair'>Pulse: <input id='tone-pulse' value='4'></div>
```
Duty in % to determine duty width. ( FamiTracker calls the pulse-width as "duty" ). And pulse, number of pulses packed in a single audio buffer in unit of frequency instead of period in order to make pure tones into audio buffer more easily to make harmonics.

Kouzeru commented on index.html:
```diff
@@ -81,28 +79,59 @@
+          <div id='sprite-left' >&larr;</div>
+          <div id='sprite-right'>&rarr;</div> 
+          <div id='sprite-vflip'>&vArr;</div>
+          <div id='sprite-hflip'>&hArr;</div>
```
For the sake of density and aesthetic using the ascii arrow icons.

Using font-size medium is intentional as the vArr hArr looks too small on firefox.

I could use another unicode icons, or just label em with only text, but idk

Kouzeru commented on index.html:
```diff
@@ -81,28 +79,59 @@
+          <div id='sprite-swap'>Swap</div>
+          <div id='sprite-or'  >OR</div>
+          <div id='sprite-and' >AND</div>
+          <div id='sprite-xor' >XOR</div>
```
Bringing back the arbitrary data bitwise blending. I need these, really. They're really useful.

Kouzeru commented on index.html:
```diff
@@ -81,28 +79,59 @@
+          <div id='audio-reverse' class='button'>Reverse</div>
+          <div id='audio-invert'  class='button'>Invert</div>
+          <div id='audio-random'  class='button'>Noise</div>
+          <div id='audio-clear'   class='button'>Clear</div>
```
The reverse, invert and clear is also needed

and relabeling the "random" into noise, because...

https://user-images.githubusercontent.com/46141631/134749922-d176ab13-4732-406b-8d4e-bc8f6b654e69.mp4

Kouzeru commented on index.html:
```diff
@@ -112,14 +141,11 @@ <h1>Tone Generator</h1>
+          <div class='pair'>Pulse: <input id='tone-pulse' value='4'></div>
+          <div id='audio-generate' class='button'>Generate</div>
         </div>
+        <center><i><div id='tone-info'></div></i></center>
```
here lies the frequency display, which consist of

(frequency of tone generator's pattern with selected pitch) ≈ (the matching musical frequency) [note key] (+offset by interval) (precise offset by interval for pitch value)

Kouzeru commented on js/tool-audio.js:
```diff
@@ -2,17 +2,20 @@
-writePattern(audioPatternEditor, emptySound)
+
+var audioPatternData = range(PATTERN_SIZE).map(_ => 0)
+writePattern(audioPatternEditor, audioPatternData)
```
the pattern data editor has it's own working array of byte (as the sprite editor) for the pattern data manipulations

Kouzeru commented on js/tool-audio.js:
```diff
@@ -43,6 +56,52 @@ function playBytes(pattern, pitch){
+		case 1: audioPreview(); break;
+		case 2: tonePreview(); break;
+	}
+})
```
plays pattern as piano key feedback upon mouse release

upon mouse press? nope, I couldn't figure it out playing properly

Kouzeru commented on js/tool-audio.js:
```diff
@@ -51,37 +110,50 @@ const audioPitch = document.getElementById('audio-pitch')
+},(x, y, draw) => {
+	writePattern(audioPatternEditor, audioPatternData)
 	updateAudio()
 })
```
for the sake of performance,
drawing on audio pattern doesn't affect the editor textbox until the mouse is released

Kouzeru commented on js/tool-sprite.js:
```diff
@@ -2,15 +2,14 @@
 **/
 
-const SPRITE_SCALE  = 20
+const SPRITE_SCALE  = 16
```
its an intentional change in order to fit more interface into the sprite editor panel

Kouzeru commented on js/tool-sprite.js:
```diff
@@ -74,7 +73,51 @@ function scrollSprite(dx, dy) {
+      return getSpritePixel(row,col,dim.cols == 16, c)
+    })
+  }));
+}
```
just following the law of symmetries, the diagonal flip (by simply swapping x and y) is used for rotations

doesn't quite work when rows and cols are different

Kouzeru commented on js/tool-sprite.js:
```diff
@@ -74,7 +73,51 @@ function scrollSprite(dx, dy) {
+  for(var i=0;i<temp.length;i++)
+    spritePixels[i] = func(spritePixels[i],temp[i]);
+  showHex(spriteEditor,spritePixels); updateSpriteEditor();
+}
```
this blend function takes lambda expression to do *sprite editor data* with *sprite blend data* operations

Kouzeru commented on js/tool-sprite.js:
```diff
@@ -118,42 +161,69 @@ function showSprite() {
+  showSprite()
+}, (x, y, draw) => {
   updateSpriteEditor()
 })
```
for the sake of performance,
drawing on sprite editor doesn't affect the editor textbox until the mouse is released

so, the optimization applied to both sprite and audio editor.

Why? I had a slow computer, hence drawing in the editor was lagging

https://user-images.githubusercontent.com/46141631/134751207-3b89539f-be8c-41ba-8b24-2c91fa87d34f.mp4

Kouzeru commented on js/util.js:
```diff
@@ -54,18 +54,26 @@ function setBit(bytes, n, v) {
+      mode
+    )
+    mode = 0;
+  }
```
drawOnCanvas() to handle mouse release

Kouzeru reviewed (commented):
Some reviews, describing some of changes

Kouzeru committed:
swapping mouse click buttons for piano key

its much more convenient to play pattern from tone generator with left click on the piano key

Kouzeru committed:
Merge branch 'JohnEarnest:gh-pages' into piano-keys

Kouzeru committed:
reducing the audio rendering overhead

by adapting the quality to the environment's sample rate; the lowpass cutoffs will be also match to each sample rates configurations

Kouzeru commented on js/shared.js:
```diff
@@ -273,7 +276,7 @@ function playPattern(soundLength,buffer,pitch=PITCH_BIAS,
 	var pos = sampleState.pos;
 
-	var quality = 8;
+	var quality = Math.ceil( 192000 / audio.sampleRate ) * 2;
```
I was noticing stuttering when setting the environment's sample rate 192000hz

The changes here is to compensate the super sampling quality according the environment's sample rate.
The lowpass' cutoff frequency kept same on every environment's sample rate configuration.

Kouzeru committed:
minor changes to sprite editor interface

to make the interface bit more dense hence it doesn't push the tabs below on some standard resolutions

Kouzeru commented:
I had tested it with three browsers (Chrome, Firefox, and Safari), It's ready!

Lemme know if there's something need to add or remove, and
I'm going to make a bit of manual of using the interface effectively.

Kouzeru committed:
an attempt for swipe support on canvas editor

Kouzeru committed:
fix touch

Kouzeru commented on js/util.js:
```diff
@@ -54,25 +54,46 @@ function setBit(bytes, n, v) {
+  target.addEventListener("touchmove",  e=>drag(touch(e)))
+  target.addEventListener("touchstart", e=>press(touch(e)))
+  target.addEventListener("touchend",   e=>release(touch(e)))
+  target.addEventListener("touchcancel",e=>release(touch(e)))
```
Touch swipe to draw on canvas. Making the sprite/audio pattern data "invert" functions became handy for undrawing.

https://user-images.githubusercontent.com/46141631/135306610-3effe5bd-3d9e-4101-b091-bde98d90dd55.mp4

Kouzeru committed:
touch fix 2
